### PR TITLE
Testflows -> Pytest

### PR DESCRIPTION
**Pull request by slvrtrn**
* Refactored most of the tests from testflows to pytest, which is much more convenient to run from the PyCharm IDE and, hopefully, clearer
* Added pytest run to the CI
* Removed all python stuff from CMakeLists
* Added a docker-compose file with ClickHouse 
* Minimal CONTRIBUTING.md (to be updated in the follow-ups)

TODO (follow-ups):
- Tests will nullable types - see FIXME in the code, `None` binds as `''` and the rendered queries have weird casts to String/LowCardinality(String). 
  
  Sample error:
  ```
  Attempt to read after eof: while converting '' to UInt8. (ATTEMPT_TO_READ_AFTER_EOF)
  ```
  Sample rendered query:
  ```
  SELECT * FROM test_sanity_simple_data_types 
  WHERE ni = _CAST(NULL, 'Nullable(String)') 
  ORDER BY i ASC, s ASC, d ASC
  ```
  instead of 
  ```
  SELECT * FROM test_sanity_simple_data_types
  WHERE ni = NULL 
  ORDER BY i ASC, s ASC, d ASC
  ```
  which is still incorrect, though. Should be actually rendered as:
  ```
  SELECT * FROM test_sanity_simple_data_types
  WHERE isNull(ni)
  ORDER BY i ASC, s ASC, d ASC
  ```
- Salvage the useful tests from the root `test.py` and `test.sh` files and make them using pytest, too
- Cleanup the `test` dir

### CODE REVIEW

**Comment by alesapin** [comment]
https://github.com/ClickHouse/clickhouse-odbc/pull/452#issuecomment-2352874343 Thank you, please replace them all with pytest.

File: test/src/e2e/test_datatypes.py [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+# TODO:
+#  Bool
+#  (U)Int128
+#  (U)Int256
+#  Decimal256
+#  DateTime64
+#  Array
+#  Tuple
+#  Map
+#  LowCardinality
```
Comment (slvrtrn) [inline test/src/e2e/test_datatypes.py:24]: Note that in the deleted [regression.py](https://github.com/ClickHouse/clickhouse-odbc/blob/2f55d016420af9dfbd503ba122c0c1549c180363/test/parameterized/regression.py), nullable tests are also marked as expected failure.
However, most of the data types tests that were "not supported" actually work, and implemented now.

File: test/docker/Dockerfile [hunk 2]
```diff
@@ -30,9 +30,6 @@ RUN apt-get update -y \
         libssl-dev \
```
Comment (slvrtrn) [inline test/docker/Dockerfile:30]: Probably should be removed entirely.

File: driver/test/datetime_it.cpp [hunk 3]
```diff
@@ -170,53 +170,53 @@ INSTANTIATE_TEST_SUITE_P(
...
+        DateTimeParams{"DateTime_TZ", "RowBinaryWithNamesAndTypes", "UTC",
             "toDateTime('2020-03-25 12:11:22', 'Asia/Kathmandu')", SQL_TYPE_TIMESTAMP,
-            "2020-03-25 09:26:22", SQL_TIMESTAMP_STRUCT{2020, 3, 25, 9, 26, 22, 0}
+            "2020-03-25 06:26:22", SQL_TIMESTAMP_STRUCT{2020, 3, 25, 6, 26, 22, 0}
         },
-        DateTimeParams{"DateTime64_9_TZ", "RowBinaryWithNamesAndTypes", "Europe/Moscow",
+        DateTimeParams{"DateTime64_9_TZ", "RowBinaryWithNamesAndTypes", "UTC",
             "toDateTime64('2020-03-25 12:11:22.123456789', 9, 'Asia/Kathmandu')", SQL_TYPE_TIMESTAMP,
-            "2020-03-25 09:26:22.123456789", SQL_TIMESTAMP_STRUCT{2020, 3, 25, 9, 26, 22, 123456789}
+            "2020-03-25 06:26:22.123456789", SQL_TIMESTAMP_STRUCT{2020, 3, 25, 6, 26, 22, 123456789}
```
Comment (slvrtrn) [inline driver/test/datetime_it.cpp:214]: adjusted for the clickhouse server running in docker with default UTC timezone

File: .github/workflows/Docker.yml [hunk 4]
```diff
@@ -1,100 +0,0 @@
-name: Build and Test - Docker
```
Comment (slvrtrn) [inline .github/workflows/Docker.yml:1]: Removed as I am not sure what is the point of this. The tests are executed without this test runner.

File: test/test.py [hunk 5]
```diff
@@ -16,6 +16,7 @@
 is_python_3 = (sys.version_info.major == 3)
 is_windows = (os.name == 'nt')
 
+
```
Comment (slvrtrn) [inline test/test.py:19]: to be revisited and rewritten in pytest, if necessary

the same with `test.sh` - it executes a lot of queries, but never asserts the output.

File: .github/workflows/Linux.yml [hunk 6]
```diff
@@ -107,7 +100,7 @@ jobs:
         -DTEST_DSN_LIST="ClickHouse DSN (ANSI);ClickHouse DSN (Unicode);ClickHouse DSN (ANSI, RBWNAT)"
 
     - name: Build
-      run: cmake --build ${{ github.workspace }}/build --config ${{ matrix.build_type }}
+      run: cmake --build ${{ github.workspace }}/build --config ${{ matrix.build_type }} --parallel $(nproc)
```
Comment (slvrtrn) [inline .github/workflows/Linux.yml:103]: speeds up the build by a few minutes

File: .github/workflows/Linux.yml [hunk 7]
```diff
@@ -122,94 +115,96 @@ jobs:
     # However, these binaries are uploaded to be available in GH's 'Actions', just in case.
     - name: Upload the artifacts
       if:  ${{ matrix.compiler == 'GCC' && matrix.odbc_provider == 'UnixODBC' && matrix.build_type == 'Release' && matrix.runtime_link == 'dynamic-runtime' && matrix.third_parties == 'bundled-third-parties' }}
-      uses: actions/upload-artifact@v3
+      uses: actions/upload-artifact@v4
```
Comment (slvrtrn) [inline .github/workflows/Linux.yml:118]: fixed outdated Node.js warning

File: .github/workflows/Linux.yml [hunk 8]
```diff
@@ -122,94 +115,96 @@ jobs:
...
       working-directory: ${{ github.workspace }}/build
       run: ctest --output-on-failure --build-config ${{ matrix.build_type }} -R '.*-ut.*'
 
-    - name: Test - Start ClickHouse server in background
+    - name: Prepare ODBC ini configs
       run: |
-        docker pull ${CLICKHOUSE_SERVER_IMAGE}
-        docker run -d --name clickhouse ${CLICKHOUSE_SERVER_IMAGE}
-        docker ps -a
-        docker stats -a --no-stream
```
Comment (slvrtrn) [inline .github/workflows/Linux.yml:139]: uses docker-compose for that, so it's more convenient locally as well.

File: .github/workflows/Linux.yml [hunk 9]
```diff
@@ -122,94 +115,96 @@ jobs:
...
+      run: |
+        pip install -r requirements.txt
+
+      # An empty `ODBCSYSINI` is required in this case to run properly with custom `ODBCINI`/`ODBCINSTINI` paths
+    - name: Test - Run Python e2e tests
+      working-directory: source/test
+      run: |
+        export ODBCSYSINI=
+        export ODBCINSTINI="${{ github.workspace }}/run/.odbcinst.ini"
+        export ODBCINI="${{ github.workspace }}/run/.odbc.ini"
```
Comment (slvrtrn) [inline .github/workflows/Linux.yml:209]: Probably should have been made a bit cleaner and extracted into some "generate env variables" step, but it works. Note the empty `ODBCSYSINI` (mentioned in the contribution guide now, too).

File: driver/test/statement_parameters_it.cpp [hunk 10]
```diff
@@ -361,8 +361,8 @@ INSTANTIATE_TEST_SUITE_P(TypeConversion, ParameterColumnRoundTripDecimalAsString
         "-12345",
         "12345.6789",
         "-12345.6789",
-        "12345.000000000000",
-        "12345.001002003000",
+        "12345",
+        "12345.001002003",
```
Comment (slvrtrn) [inline driver/test/statement_parameters_it.cpp:365]: CH does not give us the trailing zeroes in the recent versions.

File: test/parameterized/regression.py [hunk 11]
```diff
@@ -1,123 +0,0 @@
...
-        Feature(run=load("parameterized.funcvalues", test="funcvalues"), flags=TE)
-
-@TestModule
-def regression(self):
-    """The regression module for clickhouse-odbc driver.
-    """
-    Feature(run=parameterized, flags=TE)
-
-if main():
-    xfails = {
```
Comment (slvrtrn) [inline test/parameterized/regression.py:32]: Note the expected failures here.

File: test/src/e2e/test_datatypes.py [hunk 12]
```diff
@@ -0,0 +1,498 @@
...
+import decimal
+import math
+import uuid
+
+import pytest
+
+from util import pyodbc_connection, create_table, rows_as_values
+
+
+# FIXME: None is converted to an empty string (probably by PyODBC itself?)
```
Comment (slvrtrn) [inline test/src/e2e/test_datatypes.py:11]: I wonder if it's indeed the PyODBC "feature".
Maybe could try [turbodbc](https://turbodbc.readthedocs.io/en/latest/) in the future.

**Comment by slvrtrn** [comment]
It should be "mergeable," I think. Added some comments to make the review a bit easier.

File: driver/test/datetime_it.cpp [hunk 13]
```diff
@@ -170,53 +170,53 @@ INSTANTIATE_TEST_SUITE_P(
...
         }/*,
 
         // TODO: uncomment once the target ClickHouse server is 21.4+
```
Comment (mshustov) [inline driver/test/datetime_it.cpp:217]: should be uncommented now?

File: test/src/e2e/test_datatypes.py [hunk 14]
```diff
@@ -0,0 +1,498 @@
...
+#  After the fix, Nullable test cases should be re-added
+#  Sample error: Attempt to read after eof: while converting '' to UInt8. (ATTEMPT_TO_READ_AFTER_EOF)
+#
+# TODO:
```
Comment (mshustov) [inline test/src/e2e/test_datatypes.py:15]: Should we create a separate task to track these test improvements?

**Review by mshustov — APPROVED** [review]

File: driver/test/datetime_it.cpp (hunk 13 above)
Comment (slvrtrn) [reply driver/test/datetime_it.cpp:217]: it still does not work :)

**Comment by slvrtrn** [comment]
@alesapin, can you please confirm that 

1) it works on your machine 
2) CMakeLists changes are ok (there are no more Python invocations there)

After that, we can merge the branch.

**Review by alesapin — APPROVED** [review]
If it's possible let's fix timezone issue.

**Pull request merged**